#338: improve celery task handling

Toni000 opened the pull request:
Our event scheduling now uses the celery task queue to make use of it's parallelization. This means celery will start multiple  workers to do the tasks. These tasks are fully independent and performed even with over 100 patients in the same exercise, all started at once. Feel free to take a closer look, maybe we can still improve it further.

Toni000 committed:
#338: improve celery task handling

Toni000 committed:
#338: fix tests through celery eager mode (serial execution instead of parallel)

Wolkenfarmer commented on backend/dps_training_k/configuration/settings.py:
```diff
@@ -176,9 +176,7 @@
 CELERY_TASK_SERIALIZER = "json"
 CELERY_RESULT_SERIALIZER = "json"
-CELERY_TASK_TIME_LIMIT = 5 * 60  # TODO: Change if necessary
-CELERY_TASK_SOFT_TIME_LIMIT = 60  # TODO: Change if necessary
```
No reason to remove the soft time limit - the soft limit raises SoftTimeLimitExceeded which we can theoretically catch gracefully in the future. Having only the hard limit means tasks that hang will always be killed without cleanup.

Wolkenfarmer commented on backend/dps_training_k/game/tasks.py:
```diff
@@ -1,28 +1,42 @@
+    except ScheduledEvent.DoesNotExist:
+        logging.warning(f"ScheduledEvent {event_id} already deleted.")
+    except Exception as e:
+        logging.error(f"Failed to execute event {event_id}: {e}")
```
Would be good to delete the event when it failed repeatedly (same behaviour as if it succeeded) to not have our db full of failed events.

Wolkenfarmer commented on backend/dps_training_k/game/tasks.py:
```diff
@@ -1,28 +1,42 @@
+            ScheduledEvent.objects.only('id').select_for_update(skip_locked=True) # locks selected rows, skips already locked ones
+            .filter(
+                end_date__lte=settings.CURRENT_TIME(),
+                locked_by__isnull=True
```
locked_by is entirely unneccessary. It is enough to know that it was already scheduled to be executed via run_event.apply_async -> a bool value is enough.

When something is appl_asynced, it is added to a execution queue anyway where a random worker can pick it up.

Wolkenfarmer commented on backend/dps_training_k/game/tasks.py:
```diff
@@ -1,28 +1,42 @@
+        )
+        event_ids = list(claimed.values_list('id', flat=True))
+        ScheduledEvent.objects.filter(id__in=event_ids).update(
+            locked_by=worker_id
```
not necessary, see above

Wolkenfarmer commented on backend/dps_training_k/game/tasks.py:
Finally, concurrency in our django/celery setup!! Wohoo!

As our events are mostly I/O / db bound anyway, it makes sense to have multiple workers even on our 1 vCPU setup on our server. However, we will currently only spawn one celery worker there so we should overwrite the default config to also benefit on prod.

Wolkenfarmer reviewed (changes requested):
Very good performance improvement PR! It was about time we look at that - thank you for doing the work of finding this issue and presenting a fix! However, there are some things that prevent it from being fully usable for us, which I will add in subsequent commits that can be reviewed again.

Also: the commit "#338: fix tests through celery eager mode (serial execution instead of parallel)" is wrong. The reasoning why it fixes the tests is that without these changes, the events are never processed and not that they are processed in parallel otherwise. Because no celery is running, apply_async sends the events to the void. The overriden settings however cause that apply_Async effectively becomes apply and therefore the events are correctly sequentially executed.

Wolkenfarmer committed:
#338: increase celery workers per core and cap at a sensible maximum

Wolkenfarmer committed:
#338: simplify event locking

dynamic worker pickup is handled by apply_async job queuing anyway

Wolkenfarmer committed:
#338: delete repeatedly failing events

Wolkenfarmer committed:
#338: re-add celery task soft time limit

Wolkenfarmer committed:
#338: add tests for new concurrent task scheduling logic

Wolkenfarmer committed:
#338: enable celery worker count overwriting via env files

Wolkenfarmer committed:
#338: add phase change performance test

Wolkenfarmer commented:
I also now measured the performance of phase_changes due to these changes via the new performance test. I used an adapted version of it for also measuring the original main branch to see the differences. Sadly, on our server there seems to be mostly a slow down due to the 1 CPU core limitation. I still think this PR is worth it as it should improve performance in any other environment (see my local test results), but maybe we should investigate further whether we can dampen the performance hit a bit


Here are the results:

# Test results

## Server
Specs: Linode Nanode 1 GB Plan: 1 CPU Core, 25 GB Storage, 1 GB RAM
Mode: prod

### Main

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Performance Test: 100 patients
  Celery workers: 1 node(s), 1 total processes
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Setting up exercise...
Setup complete: 100 patients in 23.68s
Starting exercise (scheduling state changes for all patients)...
167 events scheduled (100 state changes, 67 pretreatment effects) in 20.28s
100 state-change events shifted to be immediately due with a buffer of 0.5.
Waiting for Celery workers to process events...
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
RESULTS
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Patients:              100
  Initial events:        100
  Setup time:            23.68s
  Scheduling time:       20.28s
  Total wall time:       3.64s
  Processing time:       3.32s
  Throughput:            30.1 events/s
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -


### PR

#### 1 celery worker

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Performance Test: 100 patients
  Celery workers: 1 node(s), 1 total processes
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Setting up exercise...
Setup complete: 100 patients in 25.03s
Starting exercise (scheduling state changes for all patients)...
172 events scheduled (100 state changes, 72 pretreatment effects) in 22.45s
100 state-change events shifted to be immediately due with a buffer of 0.5.
Waiting for Celery workers to process events...
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
RESULTS
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Patients:              100
  Initial events:        100
  Setup time:            25.03s
  Scheduling time:       22.45s
  Total wall time:       7.72s
  Processing time:       6.36s
  Throughput:            15.7 events/s
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -



#### 4 celery workers

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Performance Test: 100 patients
  Celery workers: 1 node(s), 4 total processes
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Setting up exercise...
Setup complete: 100 patients in 27.45s
Starting exercise (scheduling state changes for all patients)...
167 events scheduled (100 state changes, 67 pretreatment effects) in 19.98s
100 state-change events shifted to be immediately due with a buffer of 0.5.
Waiting for Celery workers to process events...
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
RESULTS
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Patients:              100
  Initial events:        100
  Setup time:            27.45s
  Scheduling time:       19.98s
  Total wall time:       9.27s
  Processing time:       7.79s
  Throughput:            12.8 events/s
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -




## Local
Specs: Fedora 43 Plasma Edition, Legion 5 Pro 16ITH6H with Intel Core i7 11800H in Balanced mode (16 cores)
Mode: dev

### Main

#### 1 celery worker
this would have been the default on e.g. my laptop

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Performance Test: 100 patients
  Celery workers: 1 node(s), 1 total processes
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Setting up exercise...
Setup complete: 100 patients in 18.51s
Starting exercise (scheduling state changes for all patients)...
167 events scheduled (100 state changes, 67 pretreatment effects) in 7.50s
100 state-change events shifted to be immediately due with a buffer of 0.5.
Waiting for Celery workers to process events...
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
RESULTS
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Patients:              100
  Initial events:        100
  Setup time:            18.51s
  Scheduling time:       7.50s
  Total wall time:       2.20s
  Processing time:       2.04s
  Throughput:            49.0 events/s
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -


#### 16 celery worker
this would have been the default on e.g. my laptop

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Performance Test: 100 patients
  Celery workers: 1 node(s), 16 total processes
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Setting up exercise...
Setup complete: 100 patients in 17.85s
Starting exercise (scheduling state changes for all patients)...
172 events scheduled (100 state changes, 72 pretreatment effects) in 8.03s
100 state-change events shifted to be immediately due with a buffer of 0.5.
Waiting for Celery workers to process events...
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
RESULTS
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Patients:              100
  Initial events:        100
  Setup time:            17.85s
  Scheduling time:       8.03s
  Total wall time:       2.89s
  Processing time:       2.04s
  Throughput:            48.9 events/s
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -



### PR

#### 1 celery worker

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Performance Test: 100 patients
  Celery workers: 1 node(s), 1 total processes
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Setting up exercise...
Setup complete: 100 patients in 19.25s
Starting exercise (scheduling state changes for all patients)...
164 events scheduled (100 state changes, 64 pretreatment effects) in 13.27s
100 state-change events shifted to be immediately due with a buffer of 0.5.
Waiting for Celery workers to process events...
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
RESULTS
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Patients:              100
  Initial events:        100
  Setup time:            19.25s
  Scheduling time:       13.27s
  Total wall time:       3.43s
  Processing time:       3.22s
  Throughput:            31.1 events/s
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -


#### 4 celery workers

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Performance Test: 100 patients
  Celery workers: 1 node(s), 4 total processes
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Setting up exercise...
Setup complete: 100 patients in 18.95s
Starting exercise (scheduling state changes for all patients)...
164 events scheduled (100 state changes, 64 pretreatment effects) in 13.06s
100 state-change events shifted to be immediately due with a buffer of 0.5.
Waiting for Celery workers to process events...
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
RESULTS
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Patients:              100
  Initial events:        100
  Setup time:            18.95s
  Scheduling time:       13.06s
  Total wall time:       1.35s
  Processing time:       0.94s
  Throughput:            106.9 events/s
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -


#### 32 celery workers

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Performance Test: 100 patients
  Celery workers: 1 node(s), 32 total processes
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Setting up exercise...
Setup complete: 100 patients in 18.49s
Starting exercise (scheduling state changes for all patients)...
164 events scheduled (100 state changes, 64 pretreatment effects) in 13.16s
100 state-change events shifted to be immediately due with a buffer of 0.5.
Waiting for Celery workers to process events...
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
RESULTS
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Patients:              100
  Initial events:        100
  Setup time:            18.49s
  Scheduling time:       13.16s
  Total wall time:       0.58s
  Processing time:       0.37s
  Throughput:            269.1 events/s
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -


#### 32 celery workers, 1000 patients

Note that it this point other race conditions in the code start to pop up at a frequency where the performance test run is troublesome to execute

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
Performance Test: 1000 patients
  Celery workers: 1 node(s), 32 total processes
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Setting up exercise...
Setup complete: 1000 patients in 183.22s
Starting exercise (scheduling state changes for all patients)...
1709 events scheduled (1000 state changes, 709 pretreatment effects) in 88.67s
1000 state-change events shifted to be immediately due with a buffer of 0.5.
Waiting for Celery workers to process events...
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
RESULTS
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  Patients:              1000
  Initial events:        1000
  Setup time:            183.22s
  Scheduling time:       88.67s
  Total wall time:       5.58s
  Processing time:       3.62s
  Throughput:            276.1 events/s
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Wolkenfarmer committed:
#338: optimize for one core setting

also sets celery worker scaling back to one per core as previous test results proved diminishing returns of multiple workers per core (at least in one core setting)

Wolkenfarmer commented:
Due to the observed deminishing returns of having multiple celery workers per core (at least in single-core scenarios), I reverted back to the default scaling behaviour but kept the 32 max.
This resulted only in a minor speed loss on my machine (238.1 vs 269.1 events/s).

I also optimized the code for single-core scenarios by inlining the task execution if there is only one celery worker.
This keeps the other enhancements intact (proper db locking prevent race conditions, parallel processing on multicore machines, task retry logic) and results in only a minor speed-hit compared to main (27.9 (now) vs 15.7 (old) vs 30.1 (current main) events/s).
This is imo acceptable for the other benefits.

Here is a comprehensive comparison of all test results again:
```
  ┌────────┬────────────────────────┬──────────────┬────────────────┐
  │ Branch │        Machine         │   Workers    │   Throughput   │
  ├────────┼────────────────────────┼──────────────┼────────────────┤
  │ main   │ Server (1 vCPU, prod)  │ 1 (default)  │ 30.1 events/s  │
  ├────────┼────────────────────────┼──────────────┼────────────────┤
  │ PR v1  │ Server (1 vCPU, prod)  │ 1            │ 15.7 events/s  │
  ├────────┼────────────────────────┼──────────────┼────────────────┤
  │ PR v1  │ Server (1 vCPU, prod)  │ 2            │ ~15.7 events/s │
  ├────────┼────────────────────────┼──────────────┼────────────────┤
  │ PR v1  │ Server (1 vCPU, prod)  │ 4            │ 12.8 events/s  │
  ├────────┼────────────────────────┼──────────────┼────────────────┤
->│ PR v2  │ Server (1 vCPU, prod)  │ 1 (default)  │ 27.9 events/s  │
  ├────────┼────────────────────────┼──────────────┼────────────────┤
  │ main   │ Laptop (16 cores, dev) │ 1            │ 49.0 events/s  │
  ├────────┼────────────────────────┼──────────────┼────────────────┤
  │ main   │ Laptop (16 cores, dev) │ 16 (default) │ 48.9 events/s  │
  ├────────┼────────────────────────┼──────────────┼────────────────┤
  │ PR v1  │ Laptop (16 cores, dev) │ 1            │ 31.1 events/s  │
  ├────────┼────────────────────────┼──────────────┼────────────────┤
  │ PR v1  │ Laptop (16 cores, dev) │ 4            │ 106.9 events/s │
  ├────────┼────────────────────────┼──────────────┼────────────────┤
  │ PR v1  │ Laptop (16 cores, dev) │ 32           │ 269.1 events/s │
  ├────────┼────────────────────────┼──────────────┼────────────────┤
->│ PR v2  │ Laptop (16 cores, dev) │ 16 (default) │ 238.1 events/s │
  └────────┴────────────────────────┴──────────────┴────────────────┘
```

Toni000 commented on backend/dps_training_k/deployment/django/celery/worker/start:
```diff
@@ -3,5 +3,16 @@
+fi
 
-watchgod celery.__main__.main --args -A configuration.celery_app worker -l INFO
+export CELERY_WORKER_CONCURRENCY=$CONCURRENCY
```
Why introduce the CONCURRENCY variable, when you could just use CELERY_WORKER_CONCURRENCY directly?

Toni000 commented on backend/dps_training_k/game/migrations/0007_scheduledevent_locked_by.py:
This could simply be deleted, no?

Toni000 commented on backend/dps_training_k/game/migrations/0008_remove_scheduledevent_locked_by_and_more.py:
If the previous migration is deleted, this migration would need to be updated

Toni000 commented on backend/dps_training_k/game/management/commands/performance_test.py:
```diff
@@ -0,0 +1,225 @@
+
+
+class Command(BaseCommand):
+    help = "Creates 100 patients and measures how quickly Celery processes the initial phase changes."
```
since number of patients is configurable, it shouldn't be hardcoded here.

Toni000 commented on backend/dps_training_k/game/tasks.py:
```diff
@@ -1,28 +1,51 @@
-# When this method is changed, the docker container needs to be rebuilt, because the
-# Celery-worker doesn't get the update based on our docker configuration
+# When any of these methods is changed, the docker container needs to be rebuilt, because the
+# Celery-worker doesn't get the update with our docker configuration
```
Just to keep in mind: the new deployment should fix this, but for now the comment may stay

Toni000 reviewed (commented):
Mostly looks good to me, just some minor things I'd like to see changed before we merge :)
